src/dissect/dissect.c

```diff
@@ -823,7 +841,82 @@ static int action_copy(DissectedImage *m, LoopDevice *d) {
         return 0;
 }
 
+static int action_umount(const char *path) {
+        _cleanup_free_ char *absolute = NULL;
+        dev_t devno;
+        const char *node;
+        _cleanup_(loop_device_unrefp) LoopDevice *d = NULL;
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *part;
+        int r;
+
+        r = path_make_absolute_cwd(path, &absolute);
+        if (r < 0)
+                return log_error_errno(r, "Failed to make path '%s' absolute: %m", path);
+
+        r = path_get_whole_disk(absolute, /*backing=*/ true, &devno);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find backing block device for '%s': %m", absolute);
+
+        r = sd_device_new_from_devnum(&device, 'b', devno);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create sd-device object for device %u:%u: %m",
+                                       major(devno), minor(devno));
+
+        r = sd_device_get_devname(device, &node);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get block device devname: %m");
+
+        r = loop_device_open(node, 0, /*relinquish=*/ false, &d);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open loop device '%s': %m", node);
+
+        r = loop_device_flock(d, LOCK_EX);
+        if (r < 0)
+                return log_error_errno(r, "Failed to lock loop device '%s': %m", node);
+
+        r = umount_recursive(absolute, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to unmount '%s': %m", absolute);
+
+        r = sd_device_enumerator_new(&e);
```

Comment:
i'd split the rest from here down into a separate function delete_block_device_partitions() or so, that takes the sd_device().

And I wonder if this shouldn't rather be moved inside of the loop_device_unref(), so that we can use it all over the place

Reply:
Also, a comment would be good, that explains that we do this to gurantee synchronous deletion as much as reasonably possible.